### PR TITLE
feat(i18n): bulk locale publish modal in CM edit view

### PR DESCRIPTION
**Pull request by jhoward1994**
### What does it do?

Adds bulk locale publish modal to the CM edit view

### Why is it needed?

V5 core feature

### How to test it?

Added E2E test

### Related issue(s)/PR(s)

DX-1364

TODO

- [x] Errors cases. Handle API errors and display them in the modal
- [x] Tests
- [x] ~~What tracking events should we send from this action~~ Emit event to be handled by the doc service

### CODE REVIEW

**Comment by jhoward1994** [comment]
Hey @joshuaellis this is still very very WIP still but I thought this was a good point to check in with you on the way I'm implementing this. wdyt?

File: packages/core/content-manager/admin/src/exports.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@
  * but should still export the same things.
  */
 
+export { DocumentStatus } from './pages/EditView/components/DocumentStatus';
```
Comment (joshuaellis) [inline packages/core/content-manager/admin/src/exports.ts:7]: I'd do export * just then you have the types, just incase you need them 😄

File: packages/plugins/i18n/admin/src/index.ts [hunk 2]
```diff
@@ -71,6 +75,19 @@ export default {
       return actions;
     });
 
+    contentManager.apis.addDocumentAction((actions: DocumentActionComponent[]) => {
+      const hasPublishAction = actions.find((action) => action.type === 'publish');
+      if (!hasPublishAction) {
+        return actions;
+      }
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/index.ts:82]: Are you doing this to determine if we have D&P? :)

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 3]
```diff
@@ -226,6 +228,149 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+      label: formatMessage({
+        id: 'TODO translation key',
+        defaultMessage: 'Unpublish multiple locales',
+      }),
+      icon: <Layer />,
+      position: ['panel'],
+      variant: 'secondary',
+      onClick: () => {},
+    };
+  }
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:331]: Is this in the design? If so it's a separate action...right?

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      baseStats.published.amount++;
+    } else if (status === 'modified' || status === 'draft') {
+      baseStats.draft.amount++;
+    } else {
+      baseStats.default.amount++;
+    }
+  });
+
+  return Object.values(baseStats);
+};
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:48]: could be a reduce function?

**Review by joshuaellis — COMMENTED** [review]
Overall looking good 💪🏼

File: packages/plugins/i18n/admin/src/index.ts (hunk 2 above)
Comment (jhoward1994) [reply packages/plugins/i18n/admin/src/index.ts:82]: yep

Comment (joshuaellis) [reply packages/plugins/i18n/admin/src/index.ts:82]: i'd probably just add it, and check the schema options are return `null` if it's not enabled? we add all the EE parts and then return null regardless

File: packages/core/content-manager/admin/src/hooks/useDocumentActions.ts [hunk 5]
```diff
@@ -295,6 +303,48 @@ const useDocumentActions: UseDocumentActions = () => {
     [trackUsage, publishDocument, toggleNotification, formatMessage, formatAPIError]
   );
 
+  const [publishManyDocuments] = usePublishManyDocumentsMutation();
+  const bulkPublish: IUseDocumentActs['bulkPublish'] = React.useCallback(
+    async ({ model, documentIds, params }) => {
+      try {
+        // TODO Confirm tracking events for bulk publish?
```
Comment (joshuaellis) [inline packages/core/content-manager/admin/src/hooks/useDocumentActions.ts:310]: @madhurisandbhor will know these :)

File: packages/core/content-manager/admin/src/utils/validation.ts [hunk 6]
```diff
@@ -199,7 +216,7 @@ const addRequiredValidation: ValidationFn = (attribute) => (schema) => {
     });
   }
 
-  return schema.nullable();
+  return schema?.nullable ? schema.nullable() : schema;
```
Comment (joshuaellis) [inline packages/core/content-manager/admin/src/utils/validation.ts:219]: Why would a schema not be nullable?

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+    id?: unknown;
+    defaultMessage?: unknown;
+  };
+
+  return (
+    typeof obj === 'object' &&
+    typeof possibleMessageDescriptor.id === 'string' &&
+    typeof possibleMessageDescriptor.defaultMessage === 'string'
+  );
+};
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:44]:
```suggestion
const isMessageDescriptor = (obj: unknown): obj is MessageDescriptor => {
  return (
    typeof obj === 'object' &&
    'id' in obj && typeof obj.id === 'string' &&
    "defaultMessage" in obj && typeof obj.defaultMessage === 'string'
  );
};
```

This is what TS wanted you to do :)

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+      .map(([key, value]) => {
+        if (isMessageDescriptor(value?.defaultMessage)) {
+          return `${key}: ${formatMessage(value.defaultMessage)}`;
+        } else if (isMessageDescriptor(value)) {
+          return `${key}: ${formatMessage(value)}`;
+        }
+
+        return `${key}: ${value}`;
+      })
+      .join(' ');
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:60]: I find this quite weird, how would the defaultMessage of a validation error be a MessageDescriptor?

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+  onClose,
+  validationErrors = {},
+}: BulkLocaleActionModalProps) => {
+  const { formatMessage } = useIntl();
+  const navigate = useNavigate();
+
+  const selectedRows: LocaleStatus[] = useTable(
+    'BulkLocaleActionModal',
+    (state: { selectedRows: LocaleStatus[] }) => state.selectedRows
+  );
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:152]:
```suggestion
  const selectedRows: LocaleStatus[] = useTable<LocaleStatus[]>(
    'BulkLocaleActionModal',
    (state) => state.selectedRows
  );
```

I _think_ this should work...

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+  const getFormattedCountMessage = () => {
+    const currentStatusByLocale = rows.reduce((acc: Record<string, string>, { locale, status }) => {
+      acc[locale] = status;
+      return acc;
+    }, {});
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:158]:
```suggestion
    const currentStatusByLocale = rows.reduce<Record<string, string>>((acc, { locale, status }) => {
      acc[locale] = status;
      return acc;
    }, {});
```

File: packages/plugins/i18n/admin/src/index.ts [hunk 11]
```diff
@@ -74,6 +78,14 @@ export default {
       return actions;
     });
 
+    contentManager.apis.addDocumentAction((actions: DocumentActionComponent[]) => {
+      // TODO add to the right place in the list
+      // actions.splice(indexOfPublishAction, 0, BulkPublishAction);
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/index.ts:83]: forgotten about?

File: packages/core/content-manager/admin/src/hooks/useDocumentActions.ts [hunk 12]
```diff
@@ -533,6 +583,7 @@ const useDocumentActions: UseDocumentActions = () => {
     discard,
     getDocument,
     publish,
+    bulkPublish,
```
Comment (joshuaellis) [inline packages/core/content-manager/admin/src/hooks/useDocumentActions.ts:586]: Lets call it `publishMany` I think it's better name imo

File: packages/core/content-manager/admin/src/utils/validation.ts [hunk 13]
```diff
@@ -98,11 +98,28 @@ const createYupSchema = (
...
                 [name]: transformSchema(
-                  yup.array().of(
-                    yup.object().shape({
-                      id: yup.string().required(),
-                    })
-                  )
+                  yup.lazy((value) => {
+                    if (!value) {
+                      return yup.mixed().nullable(true);
+                    } else if (Array.isArray(value)) {
```
Comment (joshuaellis) [inline packages/core/content-manager/admin/src/utils/validation.ts:104]: I don't think this will ever be the case in the admin? 🤔  Interested to hear why you added it

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+import { type MessageDescriptor, useIntl } from 'react-intl';
+import { useNavigate } from 'react-router-dom';
+import styled from 'styled-components';
+
+import { Locale } from '../../../shared/contracts/locales';
+import { getTranslation } from '../utils/getTranslation';
+
+import { LocaleStatus } from './CMHeaderActions';
+
+type status = Modules.Documents.Params.PublicationStatus.Kind | 'modified';
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx:18]:
```suggestion
type Status = Modules.Documents.Params.PublicationStatus.Kind | 'modified';
```

Types should always be capitalised 👍🏼

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 15]
```diff
@@ -226,6 +236,302 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+    return null;
+  }
+
+  if (!documentId) {
+    return null;
+  }
+
+  if (!canPublish) {
+    return null;
+  }
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:397]: I think if you can't do the action, we disable it? but if the action shouldn't be visible e.g. no D&P / no i18n then we return null 🤔 iirc....

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 16]
```diff
@@ -226,6 +236,302 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+  // This document action can be enabled given that draft and publish and i18n are
+  // enabled and we can publish the current locale.
+
+  const publish = async () => {
```
Comment (joshuaellis) [inline packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:401]: do we not do a notification for success?

File: packages/core/content-manager/admin/src/hooks/useDocumentActions.ts (hunk 12 above)
Comment (jhoward1994) [reply packages/core/content-manager/admin/src/hooks/useDocumentActions.ts:586]: 👍🏻 even though that wont match the endpoint on the BE? Or you mean to update
it there too?

File: packages/core/content-manager/admin/src/utils/validation.ts (hunk 13 above)
Comment (jhoward1994) [reply packages/core/content-manager/admin/src/utils/validation.ts:104]: Wasn't an array the only case here before this PR?

The main thing I am trying to add is

```
          } else if (typeof value === 'object') {
                      // TODO assert for specific keys such as connect, disconnect, count?
                      return yup.object();
                    } else {

```

Because otherwise the bulk locale publish modal displays errors for relation
fields (caused by the useDocument validate) when the value in the document is something like { count: 0 } 

e.g. the e2e tests were failing here
https://github.com/strapi/strapi/blob/e93e2f954ae682623665528511b8af3eb0784cb8/tests/e2e/tests/i18n/editview.spec.ts#L311

Because of the empty author relation

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx (hunk 16 above)
Comment (jhoward1994) [reply packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:401]: I think it's handled here?

https://github.com/strapi/strapi/blob/e93e2f954ae682623665528511b8af3eb0784cb8/packages/core/content-manager/admin/src/hooks/useDocumentActions.ts#L322-L328

File: packages/plugins/i18n/admin/src/index.ts (hunk 11 above)
Comment (jhoward1994) [reply packages/plugins/i18n/admin/src/index.ts:83]: Oops yes and still not sure how to put this in the right place. any ideas? I'll go back to it

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx [hunk 17]
```diff
@@ -226,6 +236,302 @@ const DeleteLocaleAction: DocumentActionComponent = ({
...
+    try {
+      const bulkPublishRes = await bulkPublishAction({
+        model,
+        documentIds: [documentId],
+        params: {
+          locale: localesToPublish,
+        },
```
[the remark this answers is not in the document]
Comment (jhoward1994) [reply packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx]: https://github.com/strapi/strapi/commit/510eea978b21a73f4cf075f16e1adaae1ddf7367

Do you mean like this?

File: packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx (hunk 8 above)
Comment (jhoward1994) [reply packages/plugins/i18n/admin/src/components/BulkLocaleActionModal.tsx]: I think this has been improved in [`9d0b637` (#20069)](https://github.com/strapi/strapi/pull/20069/commits/9d0b63715081e6cb1ba631408b7231616c379a6d)

The issue is that the `message` property isn't always a string which doesn't match the types

File: packages/core/content-manager/admin/src/hooks/useDocumentActions.ts (hunk 12 above)
Comment (joshuaellis) [reply packages/core/content-manager/admin/src/hooks/useDocumentActions.ts]: i think its fine it doesnt match the endpoint, it's clearer name for the hook's API 😄

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (joshuaellis) [reply packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx]: i do!

File: packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx (hunk 16 above)
Comment (joshuaellis) [reply packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx:396]: it is lol

File: packages/core/content-manager/admin/src/utils/validation.ts (hunk 6 above)
Comment (joshuaellis) [reply packages/core/content-manager/admin/src/utils/validation.ts:219]: i'd write a comment it's not clear and feels odd to me on first glance, someone will need to know why you've made the change :)

File: packages/core/content-manager/admin/src/utils/validation.ts [hunk 18]
```diff
@@ -327,14 +344,35 @@ const extractValuesFromYupError = (
...
+          ...(message as MessageDescriptor),
+          values: extractValuesFromYupError(currentError?.type, currentError?.params),
+        };
+      } else {
+        acc[key] = {
+          id: message,
+          defaultMessage: message,
+          values: extractValuesFromYupError(currentError?.type, currentError?.params),
+        };
+      }
```
Comment (joshuaellis) [inline packages/core/content-manager/admin/src/utils/validation.ts:375]: Alright i've just realised the confusion around all this, this function works differently to how the `Form` works managing YupValidation Errors. I must have forgotten about this, lets unify them (imo the form way is easier to understand, unless you disagree?):

https://github.com/strapi/strapi/blob/a8d58e7f7f3f8fd9e661983c7b3220878380e373/packages/core/admin/admin/src/components/Form.tsx#L230

**Review by joshuaellis — COMMENTED** [review]
Your action and modal look great, clear how it all works, some minor feedback on the inners of the CM changes :)

File: packages/plugins/i18n/package.json
Comment (joshuaellis) [inline packages/plugins/i18n/package.json:1]: you want to revert this change back to peer-deps because they're installed at the same time in an app 👍🏼 

You can see use it's exports as part of production code because it is a peer-dep – just like we do with styled-components etc.

**Review by joshuaellis — APPROVED** [review]
Need to fix the package.json for i18n, but otherwise code looks good to me, @innerdvations was going to actually QA this right?
